onnxscript/rewriter/pattern_match_context_test.py

```diff
@@ -0,0 +1,222 @@
+            
+        def identity_pattern(op, x):
+            return op.Identity(x)
+
+        def replacement(op, x):
+            return x  # Remove identity
+
+        rule = pattern.RewriteRule(
+            identity_pattern, 
+            replacement,
+            condition_function=capture_context_condition
+        )
+        
+        count = rule.apply_to_model(model)
+        self.assertEqual(count, 1)
+        self.assertIsNotNone(captured_context)
+        
+        # The properties should be read-only (no setters)
+        with self.assertRaises(AttributeError):
+            captured_context.model = None
+        with self.assertRaises(AttributeError):
+            captured_context.graph_or_function = None
+        with self.assertRaises(AttributeError):
+            captured_context.main_root_node = None
+        with self.assertRaises(AttributeError):
+            captured_context.output_values = None
+        with self.assertRaises(AttributeError):
+            captured_context.nodes = None
+
+    def test_backward_compatibility_with_no_context(self):
+        """Test that condition functions that don't use context still work."""
+        
+        model_proto = onnx.parser.parse_model(
+            """
+            <ir_version: 7, opset_import: [ "" : 17]>
+            agraph (float[N] x) => (float[N] z)
+            {
+                z = Identity(x)
+            }
+        """
+        )
+        model = ir.serde.deserialize_model(model_proto)
+        
+        # Condition function that ignores context parameter (old style)
+        def old_style_condition(context, x):
+            # This should work even though context is passed as first parameter
+            return True
+            
+        def identity_pattern(op, x):
+            return op.Identity(x)
+
+        def replacement(op, x):
+            return x
+
+        rule = pattern.RewriteRule(
+            identity_pattern, 
+            replacement,
+            condition_function=old_style_condition
+        )
+        
+        count = rule.apply_to_model(model)
+        self.assertEqual(count, 1)
+
+    def test_context_usage_in_condition_function(self):
+        """Test that PatternMatchContext can be meaningfully used in condition functions."""
+        
+        model_proto = onnx.parser.parse_model(
+            """
+            <ir_version: 7, opset_import: [ "" : 17]>
+            agraph (float[N] x, float[N] y) => (float[N] z)
+            {
+                c1 = Constant<value_float = 1.0>()
+                t1 = Div(c1, x)
+                z = Mul(t1, y)
+            }
+        """
+        )
+        model = ir.serde.deserialize_model(model_proto)
+        
+        def condition_using_context(context, x, y):
+            # Use context to check properties of the match
+            self.assertIsNotNone(context.model)
+            self.assertIsNotNone(context.graph_or_function)
+            self.assertIsNotNone(context.main_root_node)
```

Comment:
Replace this by an assertion that context.main_root_node is same as model.graph.node(2)